File: src/frontend/src/app/core/services/documento.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';
import { Documento, CreateDocumento, ValidateDocumento } from '../models';

@Injectable({
  providedIn: 'root'
})
export class DocumentoService {
  private endpoint = '/documentos';

  constructor(private api: ApiService) {}

  getAll() {
    return this.api.get<Documento[]>(this.endpoint);
  }

  getById(id: number) {
    return this.api.get<Documento>(`${this.endpoint}/${id}`);
  }

  getByCandidatoId(candidatoId: number) {
    return this.api.get<Documento[]>(`${this.endpoint}/candidato/${candidatoId}`);
  }

  upload(file: File, data: CreateDocumento) {
    const formData = new FormData();
    formData.append('arquivo', file);
    formData.append('tipo', data.tipo.toString());
    formData.append('nomeArquivo', data.nomeArquivo);
    formData.append('candidatoId', data.candidatoId.toString());
    return this.api.uploadFile<Documento>(this.endpoint, formData);
  }

  validate(id: number, data: ValidateDocumento) {
    return this.api.put<Documento>(`${this.endpoint}/${id}/validar`, data);
  }

  delete(id: number) {
    return this.api.delete<void>(`${this.endpoint}/${id}`);
  }

  getDownloadUrl(id: number): string {
    return `${this.api.getBaseUrl()}${this.endpoint}/download/${id}`;
  }

  getViewUrl(id: number): string {
    return this.api.getBaseUrl() + '/documentos/' + id + '/view';
  }

  viewDocument(id: number): Observable<Blob> {
    return this.api.getBlob(`/documentos/${id}/view`);
  }

   downloadMultiple(ids: number[]): Observable<Blob> {
     return this.api.postBlob(`/documentos/download-multiple`, ids);
   }
}

File: src/frontend/ProcessoSelecao.Blazor/Services/DocumentoService.cs
using Microsoft.AspNetCore.Components.Forms;
using ProcessoSelecao.Blazor.Models;

namespace ProcessoSelecao.Blazor.Services;

public class DocumentoService
{
    private readonly ApiService _api;
    private const string Endpoint = "documentos";

    public DocumentoService(ApiService api)
    {
        _api = api;
    }

    public virtual async Task<List<Documento>> GetAllAsync()
        => await _api.GetAsync<List<Documento>>(Endpoint) ?? new();

    public virtual async Task<Documento?> GetByIdAsync(int id)
        => await _api.GetAsync<Documento>($"{Endpoint}/{id}");

    public virtual async Task<List<Documento>> GetByCandidatoIdAsync(long candidatoId)
        => await _api.GetAsync<List<Documento>>($"{Endpoint}/candidato/{candidatoId}") ?? new();

    public virtual async Task<Documento?> UploadAsync(IBrowserFile file, CreateDocumento data)
    {
        using var content = new MultipartFormDataContent();
        using var stream = file.OpenReadStream();
        content.Add(new StreamContent(stream), "arquivo", file.Name);
        content.Add(new StringContent(data.Tipo.ToString()), "tipo");
        content.Add(new StringContent(data.NomeArquivo), "nomeArquivo");
        content.Add(new StringContent(data.CandidatoId.ToString()), "candidatoId");
        return await _api.UploadFileAsync<Documento>(Endpoint, content);
    }

    public virtual async Task<Documento?> CreateWithUrlAsync(CreateDocumentoWithUrl dto)
        => await _api.PostAsync<Documento>($"{Endpoint}/with-url", dto);

    public virtual async Task<Documento?> ValidateAsync(int id, ValidateDocumento dto)
        => await _api.PutAsync<Documento>($"{Endpoint}/{id}/validar", dto);

    public virtual async Task DeleteAsync(int id)
        => await _api.DeleteAsync($"{Endpoint}/{id}");

    public virtual string GetDownloadUrl(int id)
        => $"{_api.GetBaseUrl()}/{Endpoint}/download/{id}";

    public virtual string GetViewUrl(int id)
        => $"{_api.GetBaseUrl()}/{Endpoint}/{id}/view";

    public virtual async Task<byte[]> ViewDocumentAsync(int id)
        => await _api.GetBytesAsync($"{Endpoint}/{id}/view");

    public virtual async Task<byte[]> DownloadMultipleAsync(List<int> ids)
        => await _api.PostBytesAsync($"{Endpoint}/download-multiple", ids);
}
